### ISSUE TITLE
Migrate team authentication methods to new table

### ISSUE THREAD

**Issue by tommoor** [outline/outline#1317]
In order to support the ongoing work for additional pluggable authentication methods [here](https://github.com/outline/outline/pull/1183) – we need to move authentication details away from being stored on the `teams` table.

Some things to consider as this project starts up:

- A new table along the lines of `authentication_services`
  - `id`
  - `service` e.g. 'slack'
  - `service_id` e.g. 'T234234'
- Migrate existing data on teams to new table
- Migrate all queries to use new table
- Migrate users to add one-to-many relationship to new table

This will also unlock the ability for a single user to have multiple authentication methods.

**Comment by Bloodb0ne** [comment]
Recently became interested in the project, but because this is a authentication/security issue i wanted to give some ideas on how this could be implemented. After reading how the authentication is handled it seems like the only solution is a Polymorphic Association between the `authentication_services` and **Teams**, **Users**.

Each **User** has _Many_ **AuthenticationServices**
Each **Team** has _One_ **AuthenticationService** ( from what i read thats how the team get its name )

Additionaly a user has to be somehow marked that he was invited ( doesnt seem logical to have a "empty" AuthenticationService ).

So what we end up is a table called `authentication_services`
```
id: UUID
name: STRING
data: JSONB ( this is the replacement of  slackData, and can be used for other methods )
serviceId: STRING
authenticatableId: UUID         
authenticatableType: STRING
createdAt: TIMESTAMP ( could be useful to track when the user/team has authenticated with what method )
```

And the following fields removed from:
Users
```
service
serviceId
slackData
```
Teams
```
googleId
slackId
slackData
```

Models for User and Team should include a method that returns whether or not we authenticated with certain method, `hasAuth(<serviceName>)` ( maybe even with additional argument for the serviceId if its possible for example to authenticate multiple times with google but from a different organization ). This is needed because we do tests like `googleConnected: !!team.googleId`
 
Example Model:
```js
const AuthenticationService = sequelize.define("authentication_services", {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  name: DataTypes.STRING,
  data: DataTypes.JSONB,
  serviceID: DataTypes.STRING
});

AuthenticationService.associate = models => {
  AuthenticationService.belongsTo(models.User, {
    as: "user",
    foreignKey: "authenticatableId",
    constraints: false
  });
  AuthenticationService.belongsTo(models.Team, {
    as: "team",
    foreignKey: "authenticatableId",
    constraints: false
  });
};

export default AuthenticationService;
```

The main issue is that after this type of change it might me impossible to use the same concept of `.findOrCreate` when doing the authentication. From what i saw in the Sequelize documentation it seem unclear what will happen for example if it finds a _User_ but doesn't have that type of _Authentication_ attached to it. 
So now the logic would be searching in the _AuthenticationServices_ for Team and User and based on that creating or updating User/Team accordingly ( and associating them with  _AuthenticationServices_ ).

I hope this is of some help to you, i might finish my integration of this concept and link the commit here.

**Comment by jakoblorz** [comment]
What is the progress on this? This is kind of blocking #1183. If it is clear to me what exactly is being / will be changed, then I might be able to prepare for it during the implementation of #1183, though the interaction with this layer is what makes #1183 difficult 😅. The plugin-part will be a hopefully rigorous back-and-forth with the community. 
@tommoor @Bloodb0ne

**Comment by Bloodb0ne** [comment]
My implementation ( https://github.com/Bloodb0ne/outline/pull/1 ) is pretty rough. Its the migration,new model, some changes to models and auth methods and i decided to add tests ( for AuthenticationService ) that demonstrate how my concept is used and how it retrieves data. All other tests pass, but i could not find a way to properly mock auth flow for Google or Slack to really test them out. @jakoblorz

**Comment by almereyda** [comment]
@Bloodb0ne Would you like to point a PR from your branch at this repository, in so we can discuss technical details alongside the code? From the recent conversation in #1183 it became obvious that many people are interested in seeing this realise, why we as a community could try to come together to meet the requirements anticipated by @tommoor above, and carry out this work as much as possible already together, inviting for more feedback on the go.

From https://github.com/jakoblorz/outline/pull/1#issuecomment-709234090 by @jakoblorz (from 15.10.2020) I'm learning there is also some kind of plugin architecture anticipated, that wants to be taken into account. In https://github.com/outline/outline/issues/1135#issuecomment-575712645 by @tommoor I'm learning there already is some kind of plugin system. What's the state of this subject? Is it maybe one that's better be discussed in a separate issue then, separately and combined in one place?

There wouldn't be any technical manual that we could refer to here, why we need to deconstruct these inner workings as we progress, while including the designs of the application's authors. Given the latest contributions https://github.com/outline/outline/graphs/contributors, this would most likely be @tommoor alone. I think it would be a very fine move of the community, here demanding a change more designed after the requirements of on-premise deployments, without having to invest into the SaaS offering, if it could coordinate its forces a bit and support this very fine piece of software from the bottom up.

In short: What's missing here, maybe in bullet point or checkmark form?

**Comment by tommoor** [comment]
It seems like @Bloodb0ne's solution is close, but there are some key tweaks I would make to minimize the changes that are needed to information architecture. 

### AuthenticationProvider

After some thought I think `authentication_service` would be better represented as `authentication_provider`, and also match the terminology used in passport. So:

`authentication_providers`

- `id` - uuid
- `name` - enum eg, "slack"
- `teamId` - uuid
- `createdAt` - datetime
- `serviceId` - eg, Slack team id "T23234234"  

### UserAuthenticationProvider

One to many relationship from user -> auth providers, in a first pass each user will still only have one authentication provider (as today), but it allows for multiple in the future.

`user_authentication_providers`

- `id` - uuid
- `authenticationProviderId` - uuid
- `userId` - uuid
- `serviceId` - eg, Slack user id "U2344444"
- `createdAt` - datetime

### Migrations

We'd need a migration script to safely move data from the existing columns to these new models. For self-hosted installations this should run in ms. For the SAAS version of Outline this will probably need scheduled maintenance time to run due to the number of records.

**Comment by fooness** [comment]
Any news on this?

**Comment by solarkraft** [comment]
I am also interested, since it blocks #1183, which blocks my organization from using outline.

**Comment by VityaSchel** [comment]
any news on this? looks like it has been merged but there is still no way to authentificate with login/password

**Comment by almereyda** [comment]
You can invite users from outside your team domain via email, and they will be able to login with a magic link each time, also when coming from other email domains.
